### PR TITLE
docs: Cleanup libFuzzer Japanese Documents

### PR DESCRIPTION
**Pull request by retrage**
## Type of PR

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue

#19 

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview

#19

This PR proposes reconstructing Japanese documents for libFuzzer. The major changes in this PR are:

1. Split implementation specific part from algorithm_ja.md to implementation_ja.md.
2. Merge porting_status_ja.md to implementation_ja.md.
3. Deduplicate in-code documents with node routine descriptions and add links to corresponding source code.

## Concerns (Optional)

- [ ] Performance
- [ ] Source Code Quality

This PR removes majority of the node routine descriptions. I'm not sure it is enough to just add links to the source code.
I also reconstructed the pseudo code descriptions to make it more clear for the readers. I'm not sure the refined note is technically correct.

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [ ] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

File: docs/algorithms/libfuzzer/algorithm_ja.md [hunk 1]
```diff
@@ -92,330 +86,21 @@ while( count < total_count ) {
...
+featureとして扱われる特徴は基本的にedge coverageのedgeのインデックスとそのedgeを通過した回数が用いられる。
 
-crashedはターゲットの実行結果が正常終了以外の場合にtrueになる
+上の擬似コードでは`collect_features()`で現在のシードの実行結果`exec_result`からfeatureを収集する。`collect_features()`はfeatureが見つかるたびに`unique_feature_set`と`global_feature_freqs`を更新していく。
 
-実行結果は
+また、`collect_features()`では現在のシードの実行結果がどの程度価値があるかを表すenergyをfeatureの収集と同時に計算する。energyの計算は次の観点で計算する:
 
-* 見つかった珍しいfetureの数
+* 見つかった「珍しい」fetureの数
```
Comment (Fadis) [inline docs/algorithms/libfuzzer/algorithm_ja.md]: 元々typoしてて申し訳ないんだけど多分fetureじゃなくてfeature

File: docs/algorithms/libfuzzer/algorithm_ja.md [hunk 2]
```diff
@@ -92,330 +86,21 @@ while( count < total_count ) {
...
-
-引数で受け取った実行結果の内容をlibFuzzerのFuzzer::PrintStatusForNewUnitと同じ形式でノードが持つコールバックを使って出力する。
-
-fuzzuf::algorithm::libfuzzer::print\_status\_for\_new\_unitが担当する。
-
-## 標準的なlibFuzzerの組み方
+`collect_features()`で得られた`features`の中に新しいfeatureがある場合にはその実行結果`exec_result`とmutationで得られた入力値`mut_input`をcorpusに追加する。新しいfeatureが含まれるかどうかは`is_interesting()`が行い、`features`の中に`unique_feature_set`にある「珍しい」featureが1つ以上含まれているかで判定される。最後に次の入力値を選ぶ確率`dist`を`update_distribution()`で更新する。
 
-test/algorithms/libfuzzer/execute.cpp の中で組み立てているものがオリジナルのlibFuzzerのFuzzer::Loop()に近い処理になっている
+`update_distribution()`はenergyが高い実行結果ほど次の`select_seed()`で選ばれやすくなるように乱数の分布を更新する。この結果`select_seed()`は次のような方針でシードを選択する:
```
Comment (Fadis) [inline docs/algorithms/libfuzzer/algorithm_ja.md]: このような動きをするのはentropicモードが有効な場合に限られる。entropicモードが無効な場合energyの計算は行われず、単純に「より最近見つかった物が高い確率で選ばれる」分布が用いられる

File: docs/algorithms/libfuzzer/implementation_ja.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+ここではfuzzufでのlibFuzzerの実装について解説する。
+
+## 標準的なlibFuzzerの組み方
+
+オリジナルのlibFuzzerの`Fuzzer::Loop()`に相当するファザーは[include/fuzzuf/algorithms/libfuzzer/create.hpp](/include/fuzzuf/algorithms/libfuzzer/create.hpp)の`createRunone()`として実装されている。標準的なlibFuzzerとして使う場合にはこの関数を利用することでfuzzufでのlibFuzzer実装が利用できる。
+
+## libFuzzerを構成するHierarFlowノード
+
+HierarFlowでlibFuzzerを表現するために以下のようなノードを実装している。それぞれのノードについての詳細についてはソースコードのコメントやDoxygenで生成されたドキュメントを参照してもらいたい。
```
Comment (Fadis) [inline docs/algorithms/libfuzzer/implementation_ja.md:11]: リンク先がdoxygenのドキュメントでは無くgithub上のソースコードになっているが大丈夫か(コード上でもコメントは読めるのでこちらにリンクを貼る事自体は問題ないと思う)

File: docs/algorithms/libfuzzer/algorithm_ja.md (hunk 2 above)
Comment (retrage) [reply docs/algorithms/libfuzzer/algorithm_ja.md]: ここでの記述は正確さよりもわかりやすさを重視したいと思います。最新のLLVM libFuzzerではentropicモードはデフォルトで有効でしょうか？もしそうであれば次のような記述にしたいと思います。

```suggestion
`update_distribution()`はデフォルトでenergyが高い実行結果ほど次の`select_seed()`で選ばれやすくなるように乱数の分布を更新する。この結果`select_seed()`は次のような方針でシードを選択する:
```

File: docs/algorithms/libfuzzer/implementation_ja.md (hunk 3 above)
Comment (retrage) [reply docs/algorithms/libfuzzer/implementation_ja.md:11]: これは意図したものです。正確には同一リポジトリ内のソースコードへのリンクになっています。これはコードとドキュメントの記述の同期をとりやすくするためで、どちらかが更新された場合でもその履歴を追いやすくする意図でこのようにしています。

File: docs/algorithms/libfuzzer/algorithm_ja.md (hunk 2 above)
Comment (Fadis) [reply docs/algorithms/libfuzzer/algorithm_ja.md]: fuzzufの実装ではentropicはデフォルトで無効。LLVM 10以下ではentropicは未実装、LLVM 11ではデフォルトで無効、LLVM 12ではデフォルトで有効

File: docs/algorithms/libfuzzer/implementation_ja.md (hunk 3 above)
Comment (Fadis) [reply docs/algorithms/libfuzzer/implementation_ja.md:11]: 👍

**Review by Fadis — APPROVED** [review]

**Pull request merged**